Sources/NIO/BaseSocketChannel.swift

```diff
@@ -14,6 +14,163 @@
 
 import NIOConcurrencyHelpers
 
+private struct SocketChannelLifecycleManager {
+    // MARK: Types
+    private enum State {
+        case fresh
+        case registered
+        case activated
+        case closed
+    }
+
+    private enum Event {
+        case activate
+        case register
+        case close
+    }
+
+    // MARK: properties
+    // this is queried from the Channel, ie. must be thread-safe
+    internal let isActiveAtomic = Atomic(value: false)
+    // these are only to be accessed on the EventLoop
+    internal let channelPipeline: ChannelPipeline
+    private let eventLoop: EventLoop
+    private var currentState: State = .fresh {
+        didSet {
+            assert(self.eventLoop.inEventLoop)
+            switch (oldValue, self.currentState) {
+            case (_, .activated):
+                self.isActiveAtomic.store(true)
+            case (.activated, _):
+                self.isActiveAtomic.store(false)
+            default:
+                ()
+            }
+        }
+    }
+
+    // MARK: API
+    internal init(eventLoop: EventLoop,
+                  channelPipeline: ChannelPipeline) {
+        self.channelPipeline = channelPipeline
+        self.eventLoop = eventLoop
+    }
+
+    // this is called from Channel's deinit, so don't assert we're on the EventLoop!
+    internal var canBeDestroyed: Bool {
+        if case .closed = self.currentState {
```

Comment:
I think this can be simplified as:

```
return self.currentState == .closed
```

Reply:
thanks @normanmaurer it can indeed now. Used to have more states :)